flax/linen/normalization.py

```diff
@@ -618,3 +641,208 @@ def __call__(self, x):
+
+    Args:
+      *args: positional arguments to be passed into the call method of the
+        underlying layer instance in ``self.layer_instance``.
+      update_stats: if True, update the internal ``u`` vector and ``sigma`` value
+        after computing their updated values using power iteration. This will help
+        the power iteration method approximate the true singular value more
+        accurately over time.
+      **kwargs: keyword arguments to be passed into the call method of the
+        underlying layer instance in ``self.layer_instance``.
+
+    Returns:
+      Output of the layer using spectral normalized weights.
+    """
+
+    def layer_forward(layer_instance):
+      return layer_instance(*args, **kwargs)
+
+    return map_variables(
+        layer_forward,
+        trans_in_fn=lambda vs: jax.tree_util.tree_map_with_path(
+            functools.partial(
+                self._spectral_normalize,
+                layer_instance_name=self.layer_instance.name,
+                update_stats=update_stats,
+            ),
+            vs,
+        ),
+        init=self.is_initializing(),
+        mutable=True,
+    )(self.layer_instance)
+
+  def _spectral_normalize(self, path, vs, layer_instance_name, update_stats):
+    """Compute the largest singular value using power iteration and normalize
+    the variables ``vs`` using this value. This is intended to be a helper
+    function used in this Module's ``__call__`` method in conjunction with
+    ``nn.transforms.map_variables`` and ``jax.tree_util.tree_map_with_path``.
+
+    Args:
+      path: dict key path, used for naming the ``u`` and ``sigma`` variables
+      vs: variables to be spectral normalized
+      layer_instance_name: name of the underlying ``self.layer_instance``,
+        used for naming the ``u`` and ``sigma`` variables
+      update_stats: if True, update the ``u`` vector and ``sigma`` variables
+        after computing their updated values using power iteration. This will
+        help the power iteration method approximate the true singular value
+        more accurately over time.
+    """
+    value = jnp.asarray(vs)
+    value_shape = value.shape
+
+    # Skip and return value if input is scalar, vector or if number of power iterations is less than 1
+    if value.ndim <= 1 or self.n_steps < 1:
+      return value
+    # Handle higher-order tensors.
+    elif value.ndim > 2:
+      if self.error_on_non_matrix:
+        raise ValueError(
+            f'Input is {value.ndim}D but error_on_non_matrix is True'
+        )
+      else:
+        value = jnp.reshape(value, (-1, value.shape[-1]))
+
+    u_var_name = (
+        layer_instance_name
+        + '/'
+        + '/'.join((dict_key.key for dict_key in path[1:]))
+        + '/u'
+    )
+    u_var = self.variable(
+        'batch_stats',
+        u_var_name,
+        jax.random.normal,
+        self.make_rng('params')
+        if not self.has_variable('batch_stats', u_var_name)
```

Comment:
maybe we should allow the stats collection to be set by a constructor arg, and just defaulting to 'batch_stats'

Reply:
added